File: python/0383-ransom-note.py
class Solution:
    # O(n + m) time | O(n) space
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        if len(ransomNote) > len(magazine):
            return False
        counter = Counter(ransomNote)
        for c in magazine:
            if c not in counter:
                continue
            counter[c] -= 1
            if counter[c] == 0:
                del counter[c]
            if not counter:
                return True
        return False

File: c++/0383-ransom-note.cpp
class Solution {
public:
    // O(m + n) time | O(m) space - where m is the length of the magazine
    // and n is the length of the ransom note
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<char, int> hashmap;
        for (char c : magazine) {
            ++hashmap[c];
        }
        for (char c : ransomNote) {
            if (!hashmap[c]) {
                return false;
            }
            --hashmap[c];
        }
        return true;
    }
};